src/Datasource/EntityTrait.php

```diff
@@ -242,13 +243,39 @@ public function set($property, $value = null, array $options = [])
         $options += ['setter' => true, 'guard' => $guard];
 
         foreach ($property as $p => $value) {
+            $nested = false;
+            if (strpos($p, '.') > 0 && !array_key_exists($p, $this->_properties)) {
+                list($p, $nested) = explode('.', $p, 2);
+            }
+
             if ($options['guard'] === true && !$this->accessible($p)) {
                 continue;
             }
 
             $this->dirty($p, true);
 
-            if (!array_key_exists($p, $this->_original) &&
+            if ($nested) {
+                if (!array_key_exists($p, $this->_properties)
+                    || is_null($this->_properties[$p])
+                ) {
+                    $value = new \Cake\ORM\Entity([$nested => $value]);
```

Comment:
Here's an illustration:

```php
$e = new Entity(['a' => null]);

// If we use ArrayObjects (or arrays):
$e->set('a.b.c', true); // Internally: ::_properties['a'] = ['b' => ['c' => true]];
$e->getOriginal('a.b.c'); // Either throws an Exception or gives you all of `a`.

// If we use Entities:
$e->set('a.b.c', true); // Internally: ::_properties['a'] = new Entity(['b' => new Entity(['c' => true])]);
$e->getOriginal('a.b.c'); // Works.
```

What concerns me is being able to call `::set('a.b.c')` and have it work, but **not** being able to call `::getOriginal('a.b.c')` in all of the same cases that set() does. _Sometimes_ it will work (if a, b and c are all Entities) and sometimes it won't. But of course it won't work either way for existing properties that are already arrays, so maybe I'm being paranoid?